### PR TITLE
Server2 property running multiple servers and clients

### PR DESCRIPTION
**Pull request by UlfNorell**
Runs a single central server talking to an arbitrary number of clients and other servers.

### CODE REVIEW

File: ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs [hunk 1]
```diff
@@ -1036,14 +936,423 @@ prop_bidirectional_IO data0 data1 =
...
+    go _ [] = []
+    go t (e : es) = printf "%5s: %s" (show t') (ppEvent e) : go t' es
+      where t' = t + delay e
+
 --
 -- Utils
 --
 
+labelMe :: MonadFork m => String -> m ()
+labelMe l = myThreadId >>= (`labelThread` l)
```
Comment (coot) [inline ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs:1345]: We could use `labelThisThraed` instead.

**Review by coot — CHANGES_REQUESTED** [review]
Very nice; I will need to re-read it once more though. Some minor suggestions follow.

File: ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs [hunk 2]
```diff
@@ -447,6 +400,7 @@ withBidirectionalConnectionManager
     :: forall peerAddr socket acc req resp m a.
        ( ConnectionManagerMonad m
 
+       , acc ~ [req], resp ~ [req]
```
[the remark this answers is not in the document]
Comment (UlfNorell) [reply ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs:410]: It's used in the `MuxConnectionManager` type in the continuation and in the type signatures of `serverApplication` and `reqRespInitiatorAndResponder`. I kept the `resp` and `acc` type variables for clarity, but of course we could inline them if you think that's better.

[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs:407]: Let's leave them, it's indeed easier to read the type signature.

**Review by coot — APPROVED** [review]
LGTM

**Pull request merged**